src/core/search/search.cc

```diff
@@ -79,10 +88,57 @@ struct IndexResult {
   variant<DocVec /*owned*/, const CompressedSortedSet* /* borrowed */> value_;
 };
 
+struct ProfileBuilder {
+  string GetNodeInfo(const AstNode& node) {
+    Overloaded node_info{
+        [](monostate) -> string { return ""s; },
+        [](const AstTermNode& n) { return absl::StrCat("Term{", n.term, "}"); },
+        [](const AstRangeNode& n) { return absl::StrCat("Range{", n.lo, "<>", n.hi, "}"); },
+        [](const AstLogicalNode& n) {
+          auto op = n.op == AstLogicalNode::AND ? "and" : "or";
+          return absl::StrCat("Logical{n=", n.nodes.size(), ",o=", op, "}");
+        },
+        [](const AstTagsNode& n) { return absl::StrCat("Tags{", absl::StrJoin(n.tags, ","), "}"); },
+        [](const AstFieldNode& n) { return absl::StrCat("Field{", n.field, "}"); },
+        [](const AstKnnNode& n) { return absl::StrCat("KNN{l=", n.limit, "}"); },
+        [](const AstNegateNode& n) { return absl::StrCat("Negate{}"); },
+        [](const AstStarNode& n) { return absl::StrCat("Star{}"); },
+    };
+    return visit(node_info, static_cast<const NodeVariants&>(node));
+  }
+
+  using Tp = std::chrono::steady_clock::time_point;
+
+  Tp Start() {
+    depth_++;
+    return chrono::steady_clock::now();
+  }
+
+  void Finish(Tp start, const AstNode& node, const IndexResult& result) {
+    DCHECK_GE(depth_, 1u);
+    auto took = chrono::steady_clock::now() - start;
+    size_t micros = chrono::duration_cast<chrono::microseconds>(took).count();
+    auto descr = GetNodeInfo(node);
+    profile_.events.push_back({std::move(descr), micros, depth_ - 1, result.Size()});
+    depth_--;
+  }
+
+  AlgorithmProfile Take() {
+    reverse(profile_.events.begin(), profile_.events.end());
+    return std::move(profile_);
+  }
+
+ private:
+  size_t depth_;
+  AlgorithmProfile profile_;
+};
+
 struct BasicSearch {
   using LogicOp = AstLogicalNode::LogicOp;
 
-  BasicSearch(const FieldIndices* indices) : indices_{indices}, tmp_vec_{} {
+  BasicSearch(const FieldIndices* indices, bool profile = false) : indices_{indices}, tmp_vec_{} {
+    if (profile)
+      profile_builder_ = ProfileBuilder{};
```

Comment:
Could we expose this as a method, as opposed to as a c'tor parameter?
I realize that you use unnamed instances and directly call `Search()` on them, which will have to change if you accept this proposal, but I think that's a small price to pay :)

Reply:
ssh -i "first.pem" ubuntu@ec2-54-236-64-168.compute-1.amazonaws.com

Reply:
Ok (dunno why 😄 🤷🏻‍♂️ )